Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,191 @@
+            while endPtr >= startPtr && (whiteSpaces.contains(endPtr.pointee) || endPtr.pointee == 0) {
+                endPtr = endPtr.advanced(by: -1)
+            }
+            return self.getSlice(at: startPtr - firstPtr, length: endPtr - startPtr + 1)
+        }
+    }
+    
+    public func sliceByTrimmingWhitespaces() -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: 0, length: self.readableBytes)
+    }
+    
+    public func sliceByTrimmingWhitespaces(from start: Int = 0, length: Int) -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: start, length: length)
+    }
+    
+    public func sliceByTrimmingWhitespaces(from start: Int) -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: [" ", "\t"].map({UInt8($0.utf8CString[0])}),
+                               start: start, length: self.readableBytes - start)
+    }
+    
+    /// Compares the buffer to a case insensitive `ContiguousArray<UInt8>`.
+    ///
+    /// This `ContiguousArray` could be get from applying the `asUpperCaseContiguousUTF8UIntArray`
+    ///   property on the string protocol.
+    ///
+    /// **WARNING:** MAKE SURE THAT THE STRING YOU PASS IS UPPERCASE
+    ///
+    /// This function doesn't change the indices of the buffer
+    ///
+    /// - Parameter constant: The string constant in the form of contiguous array _IN UPPER CASE_.
+    /// - Returns: Whether the ByteBuffer contains **EXACTLY** this array or no, but by ignoring case.
+    public func compareReadingToCaseInsensitiveCString(_ constant: ContiguousArray<UInt8>) -> Bool {
+        
+        let length = self.readableBytes
+        // If available is not equal the string length itself, it can't be equal
+        if self.readableBytes != constant.count { return false }
+
+        return withVeryUnsafeBytes { pointer -> Bool in
+            var pointer = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            
+            for (idx, char) in constant.enumerated() {
+                
+                guard idx < length else { return true }
+                if (pointer.pointee & 0xdf) != char { return false }
+                pointer = pointer.advanced(by: 1)
+
+            }
+            return true
+        }
+        
+    }
+    
+    /// Compares the buffer to a case sensitive `ContiguousArray<UInt8>`.
+    ///
+    /// This `ContiguousArray` could be get from applying the `asContiguousUTF8UIntArray` property
+    /// on a string protocol.
+    ///
+    /// This function doesn't change the indices of the buffer
+    ///
+    /// - Parameter constant: The string constant in the form of contiguous array.
+    /// - Returns: Whether the ByteBuffer contains **EXACTLY** this array or no.
+    public func compareReadingToCaseSensitiveCString(_ constant: ContiguousArray<UInt8>) -> Bool {
```

Comment:
We can remove this, as it's only used in tests.

Reply:
I think it will be needed as a thing may be higher performance for framework users themselves.
We may postpone it if you want, or if `String` API itself was enough for them.

Reply:
I don't want to add APIs that we don't have concrete use-cases for, because it leads to us designing APIs that don't match the eventual needs. As a separate note, it leads to larger PRs that are harder to code review.